### PR TITLE
PXC-3872: Normal user dropping system user

### PR DESCRIPTION
**Pull request by kamil-holubicki**
https://jira.percona.com/browse/PXC-3872

Problem:
When user with SYSTEM_USER privilege is dropped by the user without
SYSTEM_USER privilege on node_1, node_2 initiates consistency voting
mechanism and self-leaves the cluster.
During the investigation I've found that not only DROP USER is affected,
but ALTER USER, RENAME USER, REVOKE ALL PRIVILEGES, REVOKE ... ON,
GRANT ... ON PROCEDURE, REVOKE ... ON PROCEDURE, GRANT <role> TO,
REVOKE <role> FROM, DROP PROCEDURE.

Cause:
Affected statements are replicated as TOI. Replication was initated
before checking if SYSTEM_USER privileges for both users allow to
proceed. As the applier thread on node_2 works in superuser context,
it is allowed to proceed with affected statement, which leads to cluster
inconsistency.

Solution:
Move initiation of TOI replication after SYSTEM_USER privileges
validation.

### CODE REVIEW

**Comment by kamil-holubicki** [comment]
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/106/

File: sql/auth/auth_internal.h [hunk 1]
```diff
@@ -204,12 +204,11 @@ int replace_routine_table(THD *thd, GRANT_NAME *grant_name, TABLE *table,
...
-#ifdef WITH_WSREP
-int open_grant_tables(THD *thd, TABLE_LIST *tables, bool *transactional_tables,
-                      const char *db = WSREP_MYSQL_DB,
-                      const char *table = NULL);
-#else
 int open_grant_tables(THD *thd, TABLE_LIST *tables, bool *transactional_tables);
+#ifdef WITH_WSREP
+bool start_toi_after_open_grant_tables(THD *thd,
+                                       const char *db = WSREP_MYSQL_DB,
+                                       const char *table = NULL);
```
Comment (venkatesh-prasad-v) [inline sql/auth/auth_internal.h]: Can we use `nullptr` here?

File: sql/sql_parse.cc [hunk 2]
```diff
@@ -5340,15 +5340,24 @@ int mysql_execute_command(THD *thd, bool first_level) {
...
-          sp_revoke_privileges(thd, db, name,
-                               lex->sql_command == SQLCOM_DROP_PROCEDURE)) {
-        push_warning(thd, Sql_condition::SL_WARNING, ER_PROC_AUTO_REVOKE_FAIL,
-                     ER_THD(thd, ER_PROC_AUTO_REVOKE_FAIL));
-        /* If this happens, an error should have been reported. */
-        goto error;
+#ifdef WITH_WSREP
+      // sp_drop_routine() returns SP_INTERNAL_ERROR in case of system user
+      // privileges check failure. In such a case we don't want to proceed with
+      // sp_revoke_privieges which will fail anyway.
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc]:
```suggestion
      // sp_revoke_privileges which will fail anyway.
```

File: sql/sql_parse.cc [hunk 3]
```diff
@@ -5535,8 +5544,7 @@ int mysql_execute_command(THD *thd, bool first_level) {
         WSREP_SYNC_WAIT(thd, WSREP_SYNC_WAIT_BEFORE_SHOW);
       }
 
-      if (lex->sql_command == SQLCOM_ALTER_USER_DEFAULT_ROLE ||
```
Comment (venkatesh-prasad-v) [inline sql/sql_parse.cc:5538]: Why this change?

File: sql/auth/sql_user_table.cc [hunk 4]
```diff
@@ -2040,6 +2014,30 @@ int open_grant_tables(THD *thd, TABLE_LIST *tables,
   return 0;
 }
 
+#ifdef WITH_WSREP
+bool start_toi_after_open_grant_tables(THD *thd, const char *db,
+                                       const char *table) {
+  /* CREATE/DROP function/procedure implicitly grant priviliges.
```
Comment (venkatesh-prasad-v) [inline sql/auth/sql_user_table.cc]:
```suggestion
  /* CREATE/DROP function/procedure implicitly grant privileges.
```

File: sql/auth/sql_user_table.cc [hunk 5]
```diff
@@ -2040,6 +2014,30 @@ int open_grant_tables(THD *thd, TABLE_LIST *tables,
...
+  Check for detail comment in respected switch handler in sql_parse.cc
+  Since the original statement is already replicated using TOI
+  sub-action of grant/revoke privilges doesn't need to get replicated. */
```
Comment (venkatesh-prasad-v) [inline sql/auth/sql_user_table.cc]:
```suggestion
  sub-action of grant/revoke privileges doesn't need to get replicated. */
```

File: sql/auth/sql_user.cc [hunk 6]
```diff
@@ -3070,6 +3087,13 @@ bool mysql_drop_user(THD *thd, List<LEX_USER> &list, bool if_exists,
       return true;
     }
 
+#ifdef WITH_WSREP
+    if (start_toi_after_open_grant_tables(thd)) {
+      commit_and_close_mysql_tables(thd);
+      return true;
+    }
+#endif
```
Comment (venkatesh-prasad-v) [inline sql/auth/sql_user.cc:3089]: Can we move this block after the validation of `utility_user`?. i.e, around line no 3120 in this file.

**Review by venkatesh-prasad-v — COMMENTED** [review]
In commit message, s/initated/initiated/

File: sql/auth/sql_user_table.cc (hunk 5 above)
Comment (kamil-holubicki) [reply sql/auth/sql_user_table.cc]: done

File: sql/auth/sql_user_table.cc (hunk 4 above)
Comment (kamil-holubicki) [reply sql/auth/sql_user_table.cc]: done

File: sql/sql_parse.cc (hunk 2 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc]: done

File: sql/auth/auth_internal.h (hunk 1 above)
Comment (kamil-holubicki) [reply sql/auth/auth_internal.h]: done

File: sql/auth/sql_user.cc (hunk 6 above)
Comment (kamil-holubicki) [reply sql/auth/sql_user.cc:3089]: I think there is no need for doing this. We check if the user we want to drop is the utility user, so it will behave the same on source and replica regardless of the user executing the query.

File: sql/auth/sql_user.cc [hunk 7]
```diff
@@ -2038,6 +2041,13 @@ bool change_password(THD *thd, LEX_USER *lex_user, const char *new_password,
       return true;
     }
 
+#ifdef WITH_WSREP
```
Comment (dutow) [inline sql/auth/sql_user.cc]: Shouldn't this be in the `if(WSREP(thd))` block? The start_toi function itself also checks it, so it should work the same - but the current diff was confusing because db_name and table_name aren't set in the else block.

File: sql/sql_parse.cc (hunk 3 above)
Comment (kamil-holubicki) [reply sql/sql_parse.cc:5538]: It is not necessary. TOI was started here and then `mysql_alter_or_clear_default_roles()` was called where TOI start attempt was made again. That 2nd attempt was skipped because already in TOI.

Comment (venkatesh-prasad-v) [reply sql/sql_parse.cc:5538]: Okay. Got it. Thanks for the explanation.

File: sql/auth/sql_user_table.cc [hunk 8]
```diff
@@ -2040,6 +2014,30 @@ int open_grant_tables(THD *thd, TABLE_LIST *tables,
   return 0;
 }
 
+#ifdef WITH_WSREP
+bool start_toi_after_open_grant_tables(THD *thd, const char *db,
```
Comment (dutow) [inline sql/auth/sql_user_table.cc:2018]: Can't this result in a deadlock with unlucky timing? (both nodes lock at the same time, and try to start TOI after that)

Comment (kamil-holubicki) [reply sql/auth/sql_user_table.cc:2018]: The sequence is:
1. Lock tables
2. Start TOI

If node_1 is between (1) and (2), and node_2 executes TOI, the operation on node_1 will fail (BF-abort).
I've added the test case for this.

File: sql/auth/sql_user.cc (hunk 7 above)
Comment (kamil-holubicki) [reply sql/auth/sql_user.cc]: You are right. It will do the same, but the flow will be easier to analyze. Good point.

File: mysql-test/suite/galera/t/pxc_drop_user.test [hunk 9]
```diff
@@ -0,0 +1,115 @@
+# Test that DROP USER executed on the user with SYSTEM_USER privilege
+# is not replicated when executing user has no SYSTEM_USER
+
+--source include/galera_cluster.inc
+--source include/count_sessions.inc
```
Comment (dutow) [inline mysql-test/suite/galera/t/pxc_drop_user.test:6]: This also needs have_debug_sync with the latest changes.

**Review by dutow — COMMENTED** [review]
I added one more comment to the test case, otherwise looks good.

File: mysql-test/suite/galera/t/pxc_drop_user.test (hunk 9 above)
Comment (kamil-holubicki) [reply mysql-test/suite/galera/t/pxc_drop_user.test:6]: added

**Review by dutow — APPROVED** [review]
